File: src/main/scala/category/greedy/_763_Solution.scala
package category.greedy

/**
  * 763. Partition Labels
  */
object _763_Solution {
  def partitionLabels(S: String): List[Int] = {
    val last = new Array[Int](26)
    for (i <- S.indices) last(S(i) - 'a') = i
    var end, anchor = 0
    var ans: List[Int] = Nil
    for (idx <- 0 until S.length) {
      end = Math.max(end, last(S(idx) - 'a'))
      if (idx == end) {ans ::= (idx - anchor + 1); anchor = idx + 1}
    }
    ans.reverse
  }
}

File: src/main/python/medium/_750_800/_763_Solution.py
# 763. Partition Labels
import collections
from typing import List


class _763_Solution:
    def partitionLabels(self, S: str) -> List[int]:
        lst = [-1] * 26
        for i in range(len(S)):
            lst[ord(S[i]) - ord('a')] = i
        pointer, ans, curr = 0, [], -1
        while pointer < len(S):
            right = pointer
            while pointer <= right:
                right = max(right, lst[ord(S[pointer]) - ord('a')])
                pointer += 1
            ans.append(right - curr)
            curr = right
        return ans


if __name__ == '__main__':
    tar = _763_Solution()
    assert tar.partitionLabels("ababcbacadefegdehijhklij") == [9, 7, 8]
